[Isabel Teo Jing Lin] iP

isabelteo opened the pull request:
# Duke

> “A plan is what, a schedule is when. It takes both a plan and a schedule to get things done.”

Combine a To-do list and a Calendar by using Duke to track **all your tasks, events and their deadlines** :+1:

Features

- [x] Able to save all your previous tasks into a database
- [x] Understands more than 8 commands 
- [x] Allows you to search for tasks by a keyword
- [ ] _Reminders (coming soon)_

Commit:
Provide a minimal project template

Let's add the following features to make this a minimal project
template that students can use to start their project.
* Gradle support, a basic build script
* a skeleton JUnit test
* a basic CheckStyle configuration
* a basic I/O test setup
* support for GitHub actions
* documentation skeletons inside the docs folder

In addition, let's remove the Gradle and JavaFX tutorials as they are
not integral to a project template.

Commit:
Add Gradle support

isabelteo committed:
Added code for Duke to echo commands and exit

isabelteo committed:
Updated Duke to level 1

isabelteo committed:
Added code for Duke to add and list (Level-2)

isabelteo committed:
Added code for Duke to add and list (Level-2)

isabelteo committed:
Merge remote-tracking branch 'duke_mygithub/master'

# Conflicts:
#	src/main/java/seedu/duke/Duke.java

isabelteo committed:
Added more comments for code that lets Duke add and list (Level-2)

isabelteo committed:
Added code to enable adding, marking and unmarking of tasks (Level-3)

isabelteo committed:
Fixed error from trying to change private final attribute and improve immutability in code. Added function to deal with cases when user marks a marked todo or unmarks an unmarked todo (Level-3)

isabelteo committed:
Tried to fix the checkstyle errors (Level-3)

isabelteo committed:
Adjusted javadoc comments to fix checkstyle errors (Level-3)

isabelteo committed:
Fixed seperator dot erros (Level-3)

isabelteo committed:
Fixed seperator dot error (Level-3)

isabelteo committed:
Fix Duke's seperator dot error

isabelteo committed:
Trying to add level-2 tag

isabelteo committed:
Merge remote-tracking branch 'duke_mygithub/master'

# Conflicts:
#	src/main/java/seedu/duke/Duke.java

isabelteo committed:
Added code to meet level-4 requirements by incorporating inheritance

isabelteo committed:
Fixed checkstyle errors

isabelteo committed:
Added code to allow for delete in level-6 by changing array to arraylist for task list

isabelteo committed:
Added code for exceptions level-5

isabelteo committed:
Fixed bug in code for exceptions level-5 and added automated inputs/outputs

isabelteo committed:
Moved exceptions into a package and fixing checkstyle errors

isabelteo committed:
Fixing checkstyle errors

isabelteo committed:
Fixing checkstyle error from curly bracket in noDateException

isabelteo committed:
Fixed error in input test file

isabelteo committed:
Fixed error in input test file again

isabelteo committed:
Added newline to end of EXPECTED.TXT

isabelteo committed:
trying to fix EXPECTED.TXT

isabelteo committed:
trying to fix automated test

isabelteo committed:
trying to fix automated test

isabelteo committed:
trying to fix automated test

isabelteo committed:
Fix exceptions and organise code into packages for better readability.

isabelteo committed:
Add code to update with every command using an absolute path

isabelteo committed:
Add code for level-7 to update to database and level-8 to search

isabelteo committed:
Fix printing problem from Storage

isabelteo committed:
Add J-unit tests

isabelteo committed:
Add Javadoc comments

tanyjnaaman commented on src/main/java/seedu/duke/chatbot/Parser.java:
```diff
@@ -0,0 +1,227 @@
+     * @throws DukeException if an incomplete command is given from user
+     */
+    Task parseForDeadline(String command) throws DukeException {
+        int indexTaskName= 9;
```
Perhaps consider leaving a comment to explain why 9, since it'll help another developer understand!

isabelteo replied:
Thank you for pointing this out! I have fixed the magic numbers in Parser.java and will take note of similar codes in other files  :+1:

tanyjnaaman commented on src/main/java/seedu/duke/chatbot/Storage.java:
```diff
@@ -0,0 +1,176 @@
+     * @return {@link TaskList} with all previous tasks added
+     * @throws DukeException if database file is not found or unable to create the {@link TaskList}
+     */
+    TaskList getOldTaskList() throws DukeException {
```
Consider breaking this method up into multiple sub methods to help with nesting and readability! Also, access permission tags like "protected" or "private" could be used if you choose to break it up!

isabelteo replied:
I edited and broke it down using one more sub-method , thanks for catching this

tanyjnaaman commented on src/main/java/seedu/duke/chatbot/Ui.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * Indicates to user that their command to delete a {@link Task} is successful.
```
Nice use of @link annotations. Super useful.

tanyjnaaman commented on src/main/java/seedu/duke/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+        super(oldDeadline.getTaskName(), done, oldDeadline.getDate());
+    }
+
+    public Deadline(String taskName, boolean doneStatus, LocalDateTime date) {
```
Perhaps consider renaming `doneStatus` to a more boolean sounding one like `isDone`. I think it's fine, but there's some room for ambiguity.

isabelteo replied:
Ah that is true, I have adjusted _doneStatus_ to _isDone_, I think it is clearer - thank you!

tanyjnaaman commented on src/main/java/seedu/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    Task(String name) {
+        this.taskName = name;
+        this.done = false;
```
Perhaps consider naming it to be more boolean sounding, like `isDone`!

isabelteo replied:
renamed! Thanks for spotting this

tanyjnaaman commented on src/main/java/seedu/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public TaskList execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        TaskList newTaskList = taskList.add(taskToAdd);
+        ui.showAddResult(newTaskList,taskToAdd);
```
Don't forget whitespace between arguments!

isabelteo replied:
edited - thank you 👍

tanyjnaaman reviewed (commented):
Honestly, your code looks great. Don't see any major issues, code quality looks good - everything is neat, well documented and tidy. Left some very minor feedback regarding maybe breaking up some methods into smaller ones, but otherwise trivial comments. Great job!

isabelteo committed:
Fix to meet coding standard.

isabelteo committed:
Added comments to magic numbers based on PR review

isabelteo committed:
Fix coding standard violations pointed out on PR review

isabelteo committed:
Set up Gradle to build and fix checkstyle errors

isabelteo committed:
Try to add GUI

xzzz3 commented on src/main/java/seedu/duke/chatbot/Parser.java:
```diff
@@ -0,0 +1,264 @@
+        // +4 away from index of "/"
+        int indexStartOfDate = dateMarkerIndex + 4;  //"/by yyyy-mm-dd hh:mm"
+        if (indexStartOfDate >= command.length()) {
+            //e.g."deadline /" is invalid ; "deadline /by " is also invalid
```
I like how you sequentially considered the possible cases of user input!

xzzz3 commented on src/main/java/seedu/duke/chatbot/Ui.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Creates a Ui object that can take in commands from the user.
+     */
+    Ui() {
```
Maybe add a public prefix to the methods? I noticed that some of your Ui methods don't have that while the rest of the methods have them.

xzzz3 commented on src/main/java/seedu/duke/task/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     * @param date for end date
+     */
+    public Deadline(String taskName, LocalDateTime date) {
+        super(taskName, false, date);
```
I like how you used inheritance from abstract class to make your code cleaner!

xzzz3 commented on src/main/java/seedu/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public String toString() {
+        String marked = this.isDone() ? "X" : " ";
+        return String.format("[%s] %s", marked, this.taskName);
```
I like this concise way of converting a task to a String!

xzzz3 commented on src/main/java/seedu/duke/task/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+    public TaskList find(String search) {
+        TaskList newTaskList = new TaskList();
+
+        for (int i = 0; i < this.getNumberOfTasks(); i++) {
```
I noticed that `i` isn't used in this loop (except for getting the task at index). A for-each loop might be more concise if you prefer that!

xzzz3 reviewed (commented):
Overall I think this PR is great. I could not find much code-related problems, as most of the code are already optimized and completed beyond the minimum iP requirements. You have considered many different potential user inputs, as seen from your extensive exception classes. It is easy to read, and I believe it will become an even more successful one down the line. Keep it up!

charmainehly commented on src/main/java/seedu/duke/chatbot/Storage.java:
```diff
@@ -0,0 +1,192 @@
+package seedu.duke.chatbot;
+
+import seedu.duke.exceptions.DukeException;
```
I like that you have already organised the different classes into separate packages, I think this will definitely help with the organisation of files as the project scales up! 😄

charmainehly commented on src/main/java/seedu/duke/task/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString()
+                + String.format(" (by: %s)",this.getFormattingDateString());
```
This is just a very minor point! I noticed a few of the toString() methods in your code might need white spaces added to follow the coding standards, perhaps you could look out for them for future commits! 😊

charmainehly commented on src/main/java/seedu/duke/task/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+    /**
+     * Used to print all tasks when command "list" is called in Duke.
+     */
+    public void printTasks() {
```
Perhaps you could consider using ` for (Task t: tasks) { ... } ` to traverse over the ArrayList<Task>? I'm not sure if it is fully applicable here, but I found it useful and a little more concise for my own method bodies! :)

charmainehly commented on src/main/java/seedu/duke/task/ToDo.java:
```diff
@@ -0,0 +1,55 @@
+package seedu.duke.task;
+
+public class ToDo extends Task {
+    private final String taskType = "T";
```
I like that you have made the different task types a fixed constant across the different classes!

charmainehly reviewed (commented):
I like the level of abstraction you have implemented, and also the way you have utilised @link and {inheritDoc} in your JavaDoc comments! I've commented just some minor areas for your consideration and review :)

isabelteo committed:
Add JavaFx Level 4

isabelteo committed:
Make launcher run from Duke rather than Main

isabelteo committed:
Fix wrong package name in built.gradle
Created a Main class to merge GUI and Duke

isabelteo committed:
Merge remote-tracking branch 'tomergetowards/branch-Level-10' into branch-Level-10

# Conflicts:
#	src/main/java/seedu/duke/chatbot/Duke.java

isabelteo committed:
Edit built.gradle

isabelteo committed:
Edit built.gradle

isabelteo committed:
Made GUI work for basic functions of Duke

isabelteo committed:
Made database path relative

isabelteo committed:
Adjust main file in built.gradle

Previously, the main class name was Duke.java.
As GUI is being used, the chatbot is now launched from Launcher.java.

Main class name in 'application' was adjusted to Launcher.java

isabelteo committed:
Add assertions to code

Problem: Hard to find bugs amongst many methods。Hard to manage regressions when changing code due to many files.

Assertions were added to make sure assumptions are valid.
Assertion message point out which file and method is causing the problem.

Including the file and method helps developer to find out the bug faster

isabelteo committed:
Merge pull request #1 from isabelteo/branch-A-Assertions

Add assertions to code

isabelteo committed:
Fix code quality

isabelteo committed:
Merge pull request #2 from isabelteo/branch-A-CodeQuality

Fix code quality

isabelteo committed:
Add extension to add notes to tasks

isabelteo committed:
Merge pull request #4 from isabelteo/branch-BCD-extension

Add extension to add notes to tasks

isabelteo committed:
Improve GUI

isabelteo committed:
Merge branch 'master' into branch-A-BetterGUI

isabelteo committed:
Merge pull request #5 from isabelteo/branch-A-BetterGUI

Improve GUI

isabelteo committed:
Fix database update for note extension.

isabelteo committed:
Fix database update for edit note and delete note.

isabelteo committed:
Enable user to add start date for Event.

isabelteo committed:
Add start date to event

isabelteo committed:
Set theme jekyll-theme-dinky

isabelteo committed:
Set theme jekyll-theme-dinky

isabelteo committed:
Set theme jekyll-theme-dinky

isabelteo committed:
Update README

isabelteo committed:
Update README

isabelteo committed:
Update README.md

isabelteo committed:
Update README

isabelteo committed:
Update README

isabelteo committed:
Update photo of UI

isabelteo committed:
Update product website main page

isabelteo committed:
Update UsersGuide.md

isabelteo committed:
Update user guide

isabelteo committed:
Add command summary

isabelteo committed:
Remove developer guide and about us

isabelteo committed:
Update README

isabelteo committed:
Upload Ui.png

isabelteo committed:
Update ReadMe

isabelteo committed:
Add credits for background image

isabelteo committed:
Edit error message to fit personality of chatbat

isabelteo committed:
Edit README.md to include user guide

isabelteo committed:
Add JavaDoc and fix code quality

isabelteo committed:
Merge branch 'master' into branch-ImproveCodeQuality

isabelteo committed:
Improve code quality and add Javadoc

Improve code quality

isabelteo committed:
Try to fix bugs to build Jar

isabelteo committed:
Fix checkstyle errors

isabelteo committed:
Fix checkstyle errors
